cmd/swapd/main.go

```
@@ -301,375 +276,153 @@ func runDaemon(c *cli.Context) error {
+		conf.SwapFactoryAddress = ethcommon.Address{}
 	} else {
 		contractAddrStr := c.String(flagContractAddress)
 		if contractAddrStr != "" {
 			if !ethcommon.IsHexAddress(contractAddrStr) {
 				return nil, fmt.Errorf("%q is not a valid contract address", contractAddrStr)
 			}
-			cfg.ContractAddress = ethcommon.HexToAddress(contractAddrStr)
+			conf.SwapFactoryAddress = ethcommon.HexToAddress(contractAddrStr)
 		}
 
-		if bytes.Equal(cfg.ContractAddress.Bytes(), ethcommon.Address{}.Bytes()) {
+		if conf.SwapFactoryAddress == (ethcommon.Address{}) {
 			return nil, fmt.Errorf("flag %q or %q is required for env=%s", flagDeploy, flagContractAddress, env)
 		}
 	}
 
+	return conf, nil
+}
+
+// validateOrDeployContracts validates or deploys the swap factory. The SwapFactoryAddress field
+// of envConf should be all zeros if deploying and its value will be replaced by the new deployed
+// contract.
+func validateOrDeployContracts(c *cli.Context, envConf *common.Config, ec extethclient.EthClient) error {
+	deploy := c.Bool(flagDeploy)
+	if deploy {
+		if envConf.SwapFactoryAddress != (ethcommon.Address{}) {
+			panic("contract address should have been zeroed when envConf was initialized")
+		}
+	}
+
 	// forwarderAddress is set only if we're deploying the swap contract, and the --forwarder-address
 	// flag is set. otherwise, if we're deploying and the flag isn't set, we also deploy the forwarder.
 	var forwarderAddress ethcommon.Address
 	forwarderAddressStr := c.String(flagForwarderAddress)
 	if deploy && forwarderAddressStr != "" {
-		ok := ethcommon.IsHexAddress(forwarderAddressStr)
-		if !ok {
-			return nil, errors.New("forwarder-address is invalid")
+		if !ethcommon.IsHexAddress(forwarderAddressStr) {
+			return errors.New("forwarder-address is invalid")
 		}
 
 		forwarderAddress = ethcommon.HexToAddress(forwarderAddressStr)
 	} else if !deploy && forwarderAddressStr != "" {
 		log.Warnf("forwarder-address is unused")
```

should this return an error?

---

I'll make it return an error. I actually thought it should return an error, but didn't know if it was written this way intentionally.

---

It returns an error now and I added a unit test.  I'll resolve this one.